Docstring additions

amshoulders opened the pull request:
This PR adds and modifies docstrings in the calcfunctions file.  More additions and changes will be coming intermittently.

jdebacker commented:
@amshoulders Thanks for this contribution!  I'll test if compile appropriately in the documentation.

jdebacker commented on taxcalc/calcfunctions.py:
```diff
@@ -547,86 +1033,188 @@ def ItemDed(e17500_capped, e18400_capped, e18500_capped, e19200_capped,
-                        capped by ItemDedCap
+    ID_Casualty_c: float
+        Ceiling on the amount of casualty expense deduction allowed (dollars)
+
```
@amshoulders You don't need a space between parameters in these lists. To save lines, it'd be better to not have them, IMO.

amshoulders replied:
Good call, thanks.  I will remove the extra lines.

amshoulders commented:
> This PR adds and modifies docstrings in the calcfunctions file. More additions and changes will be coming intermittently.

This PR is now ready for review.

MattHJensen commented:
Hi @amshoulders, thanks a lot for this PR, and sorry the ground has shifted under you right as it is ready for review. Could you fix the merge conflicts and then leave another comment when the tests are passing again?

amshoulders commented:
> Hi @amshoulders, thanks a lot for this PR, and sorry the ground has shifted under you right as it is ready for review. Could you fix the merge conflicts and then leave another comment when the tests are passing again?

Hello @MattHJensen, looks like everything is fixed now. Let me know if you have any other issues.  Thanks!

MattHJensen commented:
@amshoulders, it appears that some information is lost in this PR: 

- Distinction between tax-law parameters and taxpayer characteristics.
- Some notes that offered context but weren't directly relevant to inputs/outputs. 
- Anything else?

Any suggestions for how to keep this info in the code?

jdebacker commented:
@MattHJensen writes:

> it appears that some information is lost in this PR:

> * Distinction between tax-law parameters and taxpayer characteristics.

Previously, some (but not all) docstrings in `calcfunctions.py` did separate tax law parameters from taxpayer characteristics.  Separating sections of inputs in this way won't compile in Sphinx as they would violate the Numpy docstring formatting used throughout the Tax-Calculator project.  Additionally, while many (most?) functions have the ordering of taxpayer characteristics first, followed by tax law parameters, not all of them do.  We want the docstring to list inputs in the same order they are listed in the function arguments.

I see a couple options here:
1. Add a additional note to each input entry noting whether it is a taxpayer characteristic or a tax law parameter.  e.g.:
```
     nu13: int
         Number of dependents under 13 years old.  Taxpayer characteristic.
     elderly_dependents: int
         Number of elderly dependents age 65+ in filing unit excluding taxpayer and spouse. Taxpayer characteristic.
     earned: float
         Earned income for filing unit. Taxpayer characteristic.
     MARS: int
         Filing marital status (1=single, 2=joint, 3=separate, 4=household-head, 5=widow(er)). Taxpayer characteristic.
     ALD_Dependents_thd: list
         Maximum income to qualify for dependent care deduction. Tax law parameter.
     ALD_Dependents_hc: float
         Deduction for childcare costs haircut. Tax law parameter.
     ALD_Dependents_Child_c: float
         National weighted average cost of childcare, ceiling for available childcare deduction. Tax law parameter.
     ALD_Dependents_Elder_c: float
         Eldercare deduction ceiling. Tax law parameter.
```
2. Do not add this extra text and assume users can infer whether the function argument is a taxpayer characteristic or parameter from the descriptions provided.  Note that one can also reference the [Policy Parameters](https://taxcalc.pslmodels.org/guide/policy_params.html), [Inputs](https://taxcalc.pslmodels.org/guide/input_vars.html), and [Outputs](https://taxcalc.pslmodels.org/guide/output_vars.html) chapters of the documentation if there is confusion.  Variable names also provide clues (generally, caps used for the start of parameter names and lower case for characteristics - though there are exceptions like `MARS`).  Finally, the order of the arguments is often helpful for inference (though, as noted, there is inconsistent use of the characteristics before parameters convention).

Which would option would you prefer to see here?

MattHJensen commented:
> We want the docstring to list inputs in the same order they are listed in the function arguments.

This makes sense to me.

MattHJensen commented:
@jdebacker and @amshoulders, spot checking uncovered a few other instances of information loss. Could you confirm and recommend how to handle? 

https://github.com/PSLmodels/Tax-Calculator/pull/2558/files#diff-a87ec1055728114811ff61c0bff7b84c449561b8de892f5b4f249bc0dce89017L1485-L1491

https://github.com/PSLmodels/Tax-Calculator/pull/2558/files#diff-a87ec1055728114811ff61c0bff7b84c449561b8de892f5b4f249bc0dce89017L1601

https://github.com/PSLmodels/Tax-Calculator/pull/2558/files#diff-a87ec1055728114811ff61c0bff7b84c449561b8de892f5b4f249bc0dce89017L1678

More spot checking also confirmed my thinking that this PR is really nicely done. Thanks a great deal for your work on it @amshoulders.

jdebacker commented on taxcalc/calcfunctions.py:
```diff
@@ -28,6 +28,16 @@ def BenefitPrograms(calc):
+
+    Parameters
+    ----------
+    calc: calculator object
```
Use a capital "C" for "Calculator" object

jdebacker commented on taxcalc/calcfunctions.py:
```diff
@@ -590,86 +860,131 @@ def ItemDed(e17500_capped, e18400_capped, e18500_capped, e19200_capped,
+        Schedule A: state and local real estate taxes deductible, capped as a decimal fraction of AGI
+    e19200_capped: float
+        Schedule A: interest deduction deductible, capped as decimal fraction of AGI
+    e19800_capped: float
```
It would be useful to note that these variables are capped by the `ItemDedCap` parameter.

jdebacker commented on taxcalc/calcfunctions.py:
```diff
@@ -590,86 +860,131 @@ def ItemDed(e17500_capped, e18400_capped, e18500_capped, e19200_capped,
+    c20800: float
+        Schedule A: net limited miscellaneous deductions deducted
+    ID_ps: list
+        Itemized deduction phaseout AGI start
```
Add "(Pease)" to description to note relation to Pease Limitation -- also as in the `ID_prt` and `ID_crt` parameters as before.

jdebacker commented on taxcalc/calcfunctions.py:
```diff
@@ -1478,40 +2583,31 @@ def AmOppCreditParts(exact, e87521, num, c00100, CR_AmOppRefundable_hc,
+    exact: int
+        Whether or not to do rounding of phaseout fraction
+    e87521: float
+        Total tentative AmOppCredit amount for all students
```
Note "Form 8863, line 1"

jdebacker commented on taxcalc/calcfunctions.py:
```diff
@@ -1893,6 +3343,16 @@ def Taxes(income, MARS, tbrk_base,
+
+    Parameters
+    ----------
+    calc: calculator object
```
Use capital "C" for `Calculator` object

jdebacker commented on taxcalc/calcfunctions.py:
```diff
@@ -1921,6 +3381,16 @@ def BenefitSurtax(calc):
+
+    Parameters
+    ----------
+    calc: calculator object
```
Capital "C" for `Calculator`

jdebacker commented on taxcalc/calcfunctions.py:
```diff
@@ -1893,6 +3343,16 @@ def Taxes(income, MARS, tbrk_base,
+
+    Returns
+    -------
+    benefit
```
Say type of object that `benefit` is and give description

jdebacker commented on taxcalc/calcfunctions.py:
```diff
@@ -1942,6 +3412,16 @@ def BenefitLimitation(calc):
+
+    Parameters
+    ----------
+    calc: calculator object
```
`Calculator`

jdebacker reviewed (commented):
@amshoulders Looks pretty good - made a few small comments.

jdebacker commented on taxcalc/calcfunctions.py:
```diff
@@ -1675,13 +2836,86 @@ def NonrefundableCredits(c05800, e07240, e07260, e07300, e07400,
 
     Parameters
     ----------
-    CTC_refundable: Whether the CTC is a refundable tax credit
```
@amshoulders The `CTC_refundable` parameter needs to be added to the `Parameters` section.

jdebacker commented:
@amshoulders -- per the comment from @MattHJensen above, the first two links he provides go to lines of code that describe tax law parameters that are not parameterized (e.g., because they are hardcoded into the function). It is useful to keep these notes.

To do this, you can add a "Notes" section to the docstring.  Do this after the "Returns" section.  You can see how to format in [this style guide](https://numpydoc.readthedocs.io/en/latest/format.html) (scroll down to point 13, which describes Notes in docstrings).

amshoulders commented:
@jdebacker @MattHJensen I believe I have addressed all your comments and edits (including incorporating Notes for the referenced functions).  Let me know if there is anything else you'd like to see changed.  Thank you both for your helpful comments!

jdebacker commented:
Ok - I've found some more small issues when looking at this file in more detail.  Will post those later.  I'm having trouble compiling this to the documentation correctly - will update when I get that to work.

jdebacker commented:
The issue with most of the docstrings not compiling relates to the use of the `@iterate_jit` decorator.  See limitations on Sphinx' autodoc function [here](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directives).  The solution seems to be to use the `functools.wraps`, but that will not be done in this PR.

jdebacker commented on taxcalc/calcfunctions.py:
```diff
@@ -405,45 +658,47 @@ def AGI(ymod1, c02500, c02900, XTOT, MARS, sep, DSI, exact, nu18, taxable_ubi,
+    II_prt: float
+        Personal exemption phaseout rate
+    II_no_em_nu18: float
+        Repeal personal exemtptions for dependents under age 18       
```
Typo: exemptions

jdebacker commented on taxcalc/calcfunctions.py:
```diff
@@ -476,50 +731,65 @@ def ItemDedCap(e17500, e18400, e18500, e19200, e19800, e20100, e20400, g20500,
+    e20100: float
+        Itemizable charitalb giving: other than cash/check contributions
+    e20400: float
+        Itemizable gross (before 10% AGI disregard) causalty or theft loss
```
Typo: casualty

jdebacker commented on taxcalc/calcfunctions.py:
```diff
@@ -476,50 +731,65 @@ def ItemDedCap(e17500, e18400, e18500, e19200, e19800, e20100, e20400, g20500,
+    e20400: float
+        Itemizable gross (before 10% AGI disregard) causalty or theft loss
+    g20500: float
+        Itemizable gross (before 10% AGI disregard) causalty or theft loss
```
Typo: casualty

jdebacker commented on taxcalc/calcfunctions.py:
```diff
@@ -476,50 +731,65 @@ def ItemDedCap(e17500, e18400, e18500, e19200, e19800, e20100, e20400, g20500,
+    e20400_capped: float
+        Schedule A: gross miscellaneous deductions deductible, capped by ItemDedCap as a decimal fraction of AGI
+    g20500_capped: float
+        Schedule A: gross causalty or theft loss deductible, capped aby ItemDedCap s a decimal fraction of AGI
```
Typo: casualty

jdebacker commented on taxcalc/calcfunctions.py:
```diff
@@ -476,50 +731,65 @@ def ItemDedCap(e17500, e18400, e18500, e19200, e19800, e20100, e20400, g20500,
+    e20400_capped: float
+        Schedule A: gross miscellaneous deductions deductible, capped by ItemDedCap as a decimal fraction of AGI
+    g20500_capped: float
+        Schedule A: gross causalty or theft loss deductible, capped by ItemDedCap as a decimal fraction of AGI
```
Typo: casualty

jdebacker commented on taxcalc/calcfunctions.py:
```diff
@@ -590,86 +860,131 @@ def ItemDed(e17500_capped, e18400_capped, e18500_capped, e19200_capped,
+    c19700: float
+        Schedule A: charity contributions deducted
+    c20500: float
+        Schedule A: net causalty or theft loss deducted
```
Typo: casualty

jdebacker commented on taxcalc/calcfunctions.py:
```diff
@@ -590,86 +860,131 @@ def ItemDed(e17500_capped, e18400_capped, e18500_capped, e19200_capped,
+    e20400_capped: float
+        Schedule A: gross miscellaneous deductions deductible, capped by ItemDedCap as a decimal fraction of AGI
+    g20500_capped: float
+        Schedule A: gross causalty or theft loss deductible, capped by ItemDedCap as a decimal fraction of AGI
```
Typo: casualty

jdebacker commented on taxcalc/calcfunctions.py:
```diff
@@ -590,86 +860,131 @@ def ItemDed(e17500_capped, e18400_capped, e18500_capped, e19200_capped,
+    c19700: float
+        Schedule A: charity contributions deducted
+    c20500: float
+        Schedule A: net causalty or theft loss deducted
```
Typo: casualty

jdebacker commented:
@amshoulders I'm seeing some PEP8 styling issues:
1. Several lines exceed the recommending 79 character line lenght
2. There are trailing white spaces on a number of lines and white space on some blank lines.

I can help you set up your text editor to help identify these issues easily.  And also show you how to run a code style check (which we should probably add to the CI testing).  Ping me via email and we can set up a time to discuss.

amshoulders commented:
@jdebacker Typos and whitespaces have been removed. Thanks!

jdebacker commented:
This PR looks good to me.  Thanks @amshoulders!

MattHJensen commented:
@amshoulders, thanks very much for your contribution to Tax-Calculator! @jdebacker, thanks for the review and guidance. Merging.
